EYBlockchain/nightlite
# feat: Integration of MiMC hashing for root calculation

### Description

Mirrors the changes in this now merged Timber PR: https://github.com/EYBlockchain/timber/pull/6
This PR enables Nightlite to look out for a hash switch (HASH_TYPE) defined in nightfall (or another application) and calculates sibling paths accordingly.

Shouldn't be any breaking changes since if HASH_TYPE is not defined, Nightlite will revert to using SHA as normal. Can be tested by using the MirandaWood/mimc branch on Nightfall and adding this branch (MirandaWood/mimc/nightlite) as a yalc package.

### Key changes:

- ConcatenateThenHash in utils.js now uses mimc hashing iff HASH_TYPE is set to mimc
- Commitments are calculated using shaHash (renamed from concatenateThenHash)
- Config.js changes length of zero hex depending on the HASH_TYPE (sha uses 27 bytes, mimc 32)

- File structure in setup reflects westlad/kyc2 branch: 
- Functions are abstracted out in 'common' folder
- .zok files using MiMC hashing are kept in 'mimc' folder, where generateZokratesFiles.js looks iff HASH_TYPE is set to mimc (it ignores the common and mimc folders otherwise)

- Consolidation proof! A new action which takes 20 fungible commitments owned by the same user and consolidates the transfer of them. Only possible to run using mimc due to the calculation of 20 merkle tree paths. Vks, erc20.js, generateZokratesFiles.js, and setup files in Nightfall/MirandaWood/mimc have been changed accordingly.

### QA Instructions

This PR should not have any breaking changes, as the only differences rely on the user setting HASH_TYPE to mimc. If it's not set to anything, then SHA-256 is used as normal. 

The Nightfall/MirandaWood/mimc branch includes tests which run on this branch. Before merge, you would need to add this branch as a yalc package to run them. 

### Checklist



- [ x ] I have reviewed the code changes myself
- [ x ] I have removed unused code (e.g., `console.logs`, commented out blocks, etc.)
- [ x ] I have made all necessary changes to the documentation.
- [ x ] I have added tests to cover my changes. (In Nightfall)
- [ x ] All new and existing tests passed.
- [ x ] Any dependent changes have been merged and published.
